### PR TITLE
fix: Power On Self Test fixes for PFD/ND/MFD

### LINKED ISSUE #945: Self-test only when unpowered over 10 seconds
**Mod Version**
Master (0.3)

**Describe the bug**
When reducing the brightness on the following switches to zero:
PFD, ND, Upper ECAM, Lower ECAM, right PFD, right ND.

and then raising the brightness (thereby switching the screens back on), the screens switch to a self test mode immediately. In addition, reducing (for example) the right ND, will also swap the left ND screen to auto test (and vice versa).

**To Reproduce**
1. Reduce brightness on any screen to minimum so that it switches off.
2. Increase brightness on same screen to above minimum, and the screen goes into self test mode, along with corresponding screen on the opposite side of the cockpit.

**Expected behavior**
Lowering these switches to off and then increasing the brightness knob (switching them back on) should only enter self test mode once the screen is off for 10 seconds. If less than 10 seconds, the screen should just be as it was prior to being switched off.

**Actual behaviour**
Turning the screen brightness down for any screen (ND, PFD, Upper / Lower ECAM, right ND, right PDF) switches them off, and upon switching them back on they enter self test mode immediately.

### PR DESCRIPTION
**Pull request by 2hwk**
Fixes #945

## Summary of Changes

- [x] When powered up and ND brightness knob is turned off,  PFD now waits for 10 seconds before Power On Self-Test is restarted
- [x] When powered up and ND brightness knob is turned off,  MFD now waits for 10 seconds before Power On Self-Test is restarted
- [x] When  powered up and Lower ECAM brightness knob is turned off, Lower ECAM now waits for 10 seconds before Power On Self-Test is restarted
- [x] When powered up and Upper ECAM brightness knob is turned off, Lower ECAM now waits for 10 seconds before Power On Self-Test is restarted

## References
https://discordapp.com/channels/738864299392630914/754851312306487326/758257387432968222
See #945 

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Power on test for each display (PFD/MFD/Lower ECAM/Upper ECAM):

- Power-up Aircraft
  - Observe and wait for Power On Self-Test (POST) for Displays to Complete
  - Turn Brightness knob off for less than 10 seconds, then turn display back on
    - Display should continue to display relevant information.
  - Turn Brightness knob off for at least 10 seconds, then turn display back on.
    - Display should now display Self-Test
  
For each display, testing by switching each display in turn to various on/off states:
 - Power-off Aircraft
   - Displays should stop displaying information and go blank
 - Power-on aircraft again
   - Display should begin self-test as expected

Displays should work normally as per normal flight operations in every case in flight.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by 2hwk** [comment]
Perf fix PR has merge conflicts, will fix and retest

**Comment by 2hwk** [comment]
Should be good to go

**Comment by ErickSharp** [comment]
Category: QA Tester
Name: Erick [Z-6]#6484
Date of testing: Current date 17/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3068
Object of test: Power on self-test fixes for cockpit display panels.

Testing technique: Followed recommended testing steps in PR description. 
Steps you've made to test the Pull Request:
1. Spawn at Madeira cold and dark and start APU. 
2. Displays come to life with self-test in progress indication. 
3. Once the displays are alive, turn off the brightness for less than ten seconds and turn it up again.
4. Displays come back on without reverting to the self-test screen.
5. Turn displays off for ten seconds and then turn them back on. 
6. Displays come on with the self-test screen now on display.

Flight notes (If required):

Issues (If Any): None
Overall Rating: Awesome!
Conclusions of the testing: Displays now don't revert to the self-test screen after turning them off for less than ten seconds, and go to the self-test screen when appropriate after the ten seconds. Overall, the PR did a good job, and it's ready for merger!

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -37,7 +37,7 @@
 1. [FLIGHTMODEL] Fixed structural load factor limitations - @donstim (donbikes#4084)
 1. [ECAM] Improved Ecam Memo TO Flaps and TO Config. Allows for Config 3 TO - @MisterChocker (Leon)
 1. [DISPLAYS] Optimize performance by throttling instrument updates - @XSJoJo (Jorim Jaggi)
-
+1. [DISPLAYS] Fixed POST repeat when displays are switched off for <10 seconds - @2hwk (2Cas#1022)
```
Comment (beheh) [inline .github/CHANGELOG.md:40]: Don't drop the new line here.

**Comment by beheh** [comment]
You could use one of the NXLogic nodes from https://github.com/flybywiresim/a32nx/blob/master/A32NX/html_ui/Pages/A32NX_Utils/NXLogic.js#L56, rather than implementing the countdown again and again everywhere.

An example could be the G-Force display on the lower ECAM:
https://github.com/flybywiresim/a32nx/blob/615e9af4935bfe264d1c7b49d225ba4ba6e56a52/A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/EICAS_Common.js#L70-L96

**Comment by sidnov** [comment]
Category: QA Tester
Name: sidnov #8337
Date of testing: 18/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3068
Tier of testing: 2
Object of test: POST fixes for cockpit display.

Testing technique: Tryout

Steps you've made to test the Pull Request:

1. Spawned at VABB, C/D
2. Turned on battery, external power.
3. Observed POST on the cockpit displays
4. Turned off each of the knobs and turned on within 10 seconds. Displayed information without showing POST.
5. Turned off the knobs, waited for 15 seconds and restarted display. Observed POST showing up each time.
6. Turned off aircraft entirely (de-powered external power and batteries), and restarted for proper POST messages. Bug observed in some cases.

Bug Found: PFD on the First Officer's side randomly loaded without POST. Attached video.

Media:
1.<iframe src="https://drive.google.com/file/d/1D3kFj20hXzYn5_VV4mgsaUQPJT6_p8Jq/preview" width="640" height="480"></iframe>

2.<iframe src="https://drive.google.com/file/d/11hCg1rDQ4cxHVjAPZJgEJLjV9WH-QuyH/preview" width="640" height="480"></iframe>

Issues (If Any): Bug mentioned above. Apart from this, whenever restarting displays after 10 seconds, the actual information shows up for half a second before reverting to the POST screen. Not sure if this is intentional/needs improvement. So have attached another clip for reference.

Steps to achieve issue: 
1. Turn on all the knobs by 1-2 notches.
2. Confirm displays show information
3. Turn off batteries, ext. power. Wait for aircraft to go C/D completely
4. Restart ext. power. Video attached for reference.

Overall Rating: Needs improvement
Conclusions of the testing: A bug was observed. Apart from this, a small visual improvement may be required when restarting displays after 10 seconds.

**Comment by ChruutvoLuzi** [comment]
Category: QA Tester
Name: ChruutvoLuzi | SD-PB.ch#8902
Date of testing: 18.1.2021
Version of the sim: 1.12.13.0
PR Tested: #3068 
Tier of testing: 2
Changes to observe:
PFD, ND and ECAM self test only after 10s off


Testing technique: TryOut
Steps you've made to test the Pull Request:
1. Spawned C&D
2. Turned on batterys and External power
3. Tested all the screens multiple times (off 9s= no self test, off 11s= self test
4. Turned off the Aircraft
5. Turned the Aircraft on again
6.  Checked Screens again for self test

Flight notes (When done a full flight):

Media:

https://cdn.discordapp.com/attachments/314015589595086848/800836011760287764/Microsoft_Flight_Simulator_2021.01.18_-_22.10.22.01.mp4

Issues (If Any):
Sometimes, as already mentioned above, the normal, non testing state will appear for a very short amount of time
Steps to achieve the issue
turn the off and wait more than 10s, then turn it back on.
repeat multiple times to get it (rather rare)

Overall Rating: To be improved
Conclusions of the testing: Another nice little fix, still needs some polishing with the issue described above

**Comment by BritishAvgeek** [comment]
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: DD/MM/YYYY
Version of the sim: 1.12.13.0
PR Tested: #3068 
Tier of testing: 2
Changes to observe: Power On Self Test modifcations

Testing technique: Spawned C&D at LFMN to test
Steps you've made to test the Pull Request:
1. Spawned C&D - Bat 1 and 2 on, External power on.
2.  Screens self tested as expected
3.  Cycled screen brightness to off for each system, noting the above issues and I was able to replicate that too 
4. Recycled brightness dials off, and back on - display went straight on as described above on a few occasions

Media: N/A - the media already added above, covers everything that needs to be shown for the slight improvement needed

Issues (If Any): As described above, when restarting the displays, it can trigger the info to instantly display as opposed to cycling back into selftest
Steps to achieve the issue
Turn the display brightness to off, and on again. Same steps as Review 1 and 2 above

Overall Rating: To be improved
Conclusions of the testing: Great feature but as explained by the other tests, this needs a little modification to rectify the minor problems found. Initial C&D screens on for me worked well however, self test was as described in the PR - bug only appeared when turning screens off and back on again (which I wouldn't do in-flight anyway)

**Comment by Benjozork** [comment]
Gonna wait to resume testing until bug resolved, as all reports are saying the same thing.

**Comment by 2hwk** [comment]
I've identified the root cause - when reading from the L:ACPowerStateChange SimVar which was created after the display optimisation change was merged - this occurs as the FO's PFD now updates at a reduced rate, it is possible for it to not update when this LVar is changed, causing it to miss the cue to reset the self-test if a change in power state is noticed.

**Comment by St54Kevin** [comment]
This does not have Checks

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -42,7 +42,7 @@
 1. [SOUND] Reworked engine spool up and startup sounds - @hotshotp (Boris)
 1. [SOUND] Reworked pack sounds - @hotshotp (Boris)
 1. [SOUND] Reworked touchdown sounds - @hotshotp (Boris)
-
+1. [DISPLAYS] Fixed POST repeat when displays are switched off for <10 seconds - @2hwk (2Cas#1022)
```
Comment (MisterChocker) [inline .github/CHANGELOG.md:45]:
```suggestion
1. [DISPLAYS] Fixed POST repeat when displays are switched off for <10 seconds - @2hwk (2Cas#1022)
```
```suggestion
1. [DISPLAYS] Fixed POST repeat when displays are switched off for <10 seconds - @2hwk (2Cas#1022)

```

**Review by MisterChocker — CHANGES_REQUESTED** [review]

**Comment by sidnov** [comment]
Category: QA Tester
Name: sidnov #8337
Date of testing: 29/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3068
Tier of testing: 2
Object of test: POST fixes for cockpit display.

Testing technique: Tryout

Steps you've made to test the Pull Request:

1. Spawned at VABB, C/D
2. Turned on battery, external power.
3. Observed POST on the cockpit displays
4. Turned off each of the knobs and turned on within 10 seconds. Displayed information without showing POST.
5. Turned off the knobs, waited for 15 seconds and restarted display. Observed POST showing up each time.
6. Turned off aircraft entirely (de-powered external power and batteries), waited for 5 seconds before restarting.
7. Repeated same testing steps


Issues (If Any): None

Overall Rating: Good

Conclusions of the testing: Bugs observed in the previous version of the PR have now been fixed. Good to go!

**Comment by ChruutvoLuzi** [comment]
Category: QA Tester
Name: ChruutvoLuzi | SD-PB.ch#8902
Date of testing: 29.1.2021
Version of the sim: 1.12.13.0
PR Tested: #3068 
Tier of testing: 2
Changes to observe: fixed Power On Self test


Testing technique: TryOut
Steps you've made to test the Pull Request:
1. Spawned at LSZH
2. Turned on and off the Displays a few times with less than 10s and a few times with more than 10s time elapsing

Issues (If Any): None

Overall Rating: Good
Conclusions of the testing: Works as described now, i did not come across the previous issues such as the normal state briefly flashing before the test screen

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 3/2/21
Version of the sim: 1.12.13.0
PR Tested: #3068 
Tier of testing: 2
Changes to observe:
self-test for displays

Testing technique: Loaded in different situations and looked at self-testing
Steps you've made to test the Pull Request:
1. Loaded in C&D
2. Loaded in on runway
3. Loaded in mid-air
4. Loaded in Approach 

Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.

https://user-images.githubusercontent.com/70103461/106804292-21cce580-665d-11eb-92e4-483ded0eabbf.mp4

image

Issues (If Any): 
When turning the power off and on again using external power continuously it eventually bugs out and all the screens are on without needing to self-test 

When spawning in on the runway or in the air the nd and pfd for both captain and FO are in self-test for about 5 seconds while the ECAMs are both operating normally

I also found that when on apu gen or engine gens, when you turn those off the displays stay on until batteries are also turned off but I think that could be fixed on another pr

Steps to achieve the issue

Overall Rating: To be improved
Conclusions of the testing: A few small bugs but overall a good pr just need to fix the issues listed above

**Comment by wpine215** [comment]
Any update on this PR?

**Pull request closed without merging**